Simple Project Managment Application

mkurkar opened the pull request:
This pull request introduces a major refactor and feature expansion to the Todo application. The most significant changes are the introduction of a new `Task` model to group todos, a migration of the codebase into a `backend/` directory structure, and the addition of new admin, model, and serializer logic to support tasks and enhance todo management. The admin interface is improved for both models, and the serializers are updated for better API support. Legacy code is removed from the old `app/` directory.

**Model and Data Structure Enhancements:**

* Introduced a new `Task` model to group todos, with appropriate fields, relationships, and meta options. The `Todo` model is updated to reference `Task` via a foreign key, and its `user` field is changed from a many-to-many to a foreign key. Default tags logic is centralized. (`backend/app/models/task_model.py` [[1]](diffhunk://#diff-220e758036147311e295044dfb1b3ec515eb7f1b8268b41193d515766d92cb67R1-R51) `backend/app/models/todo_model.py` [[2]](diffhunk://#diff-2735faa73813f23bc24d7b74adc306528ce1de908878cffebd87d52d2460d70cR11-R14) [[3]](diffhunk://#diff-2735faa73813f23bc24d7b74adc306528ce1de908878cffebd87d52d2460d70cL32-R54) migration files [[4]](diffhunk://#diff-cda09487c8582da9abc936a6098497449ad09bf240710efb3a5a54ecde3ae9faR1-R43) [[5]](diffhunk://#diff-49f4198cd8295fd5527fd8653da7490ed234beaf75b695b6e76317ba24532094R1-R34)
* Added and updated migration files to reflect the new models and relationships, including the removal of the old many-to-many fields and the addition of foreign keys. (`backend/app/migrations/0004_alter_todo_options_alter_todo_tags_task.py` [[1]](diffhunk://#diff-cda09487c8582da9abc936a6098497449ad09bf240710efb3a5a54ecde3ae9faR1-R43) `backend/app/migrations/0005_remove_task_todos_todo_task_remove_todo_user_and_more.py` [[2]](diffhunk://#diff-49f4198cd8295fd5527fd8653da7490ed234beaf75b695b6e76317ba24532094R1-R34)

**Admin Interface Improvements:**

* Enhanced the Django admin by adding inline editing of todos within tasks, new admin classes for both `Task` and `Todo`, and improved filtering and display logic. (`backend/app/admin.py` [backend/app/admin.pyR1-R101](diffhunk://#diff-2a14ce665bd3172dbb784031c4979c9f075970066000e6bba4ccdc7f76223eabR1-R101))

**API and Serialization Updates:**

* Added new serializers for both `Task` and `Todo`, supporting nested todos, user filtering, and enhanced representation logic. (`backend/app/serializers/task_serializer.py` [[1]](diffhunk://#diff-e09b2a63ae073d439f7507825a7ac6f513ed0c2cf02fbae8604884cd2694aa5eR1-R86) `backend/app/serializers/todo_serializer.py` [[2]](diffhunk://#diff-956201e55ba7ffbcb908f36861be343ec6ea310d5a02b4bab445ee3c9e8e2aaaR1-R34) `backend/app/serializers/__init__.py` [[3]](diffhunk://#diff-bf1009e8224722ad045bbfaab241899b1d2897f61da944fcf124a9022c6d8b12R1-R2)
* Updated model imports to use the new modular structure. (`backend/app/models/__init__.py` [backend/app/models/__init__.pyR1-R2](diffhunk://#diff-d5870c67917db2ab2c21aa4a6553eaf5b9b2d105f1c46623461fff72e7c59ae6R1-R2))

**Cleanup and Directory Restructuring:**

* Removed legacy code from the old `app/` directory, including admin, serializer, and view logic, and migrated model code to the new modular structure under `backend/app/`. (`app/admin.py` [[1]](diffhunk://#diff-2e6f3da0a0e6f1e7448e06ec253a41912c5ce0841f76e1e43f11d1b6c8d3a24bL1-L42) `app/serializers.py` [[2]](diffhunk://#diff-56f9ad9c173c4d6c28180bc39da60697bd1bcbdebf3211193d42b8f2c9c30ec7L1-L29) `app/views.py` [[3]](diffhunk://#diff-0c2bad94cd303c5944a38a94f62a63355b8bed2e102dc87c0f44de95cf135b8aL1-L74) `app/models.py` → `backend/app/models/todo_model.py` [[4]](diffhunk://#diff-2735faa73813f23bc24d7b74adc306528ce1de908878cffebd87d52d2460d70cL2-R2)

These changes lay the groundwork for more robust task and todo management, a cleaner codebase, and improved API and admin experiences.

mkurkar committed:
add Task model and update Todo model options for improved structure

mkurkar committed:
Adds comprehensive Task admin interface with permissions

Implements full admin functionality for Task model including inline todo management, user-based filtering, and permission controls.

Enhances security by restricting non-superuser access to only their own tasks and automatically assigning current user ownership.

Removes obsolete model and serializer files as part of codebase restructuring.

mkurkar committed:
move the django in to backend folder
and start new react project in frontend folder

mkurkar committed:
Update package dependencies and add devDependencies for Tailwind CSS

mkurkar committed:
Remove unused models and views for cleaner project structure

mkurkar committed:
Refactor Task and Todo models for improved relationships and clarity

mkurkar committed:
Enhance serializers and views for Task and Todo models with user-specific data handling

mkurkar committed:
Refactor Todo and Task admin configurations, update migrations, and enhance authentication settings

mkurkar committed:
feat: Implement task and todo management features

- Added new dependencies for Material UI and Emotion for styling.
- Refactored App component to include routing for Login and Dashboard pages.
- Created Sidebar component for navigation between tasks and todos.
- Developed TaskForm and TaskList components for task creation and display.
- Implemented TodoForm and TodoList components for todo management.
- Added TaskDetailView for viewing and managing todos associated with a specific task.
- Integrated API service for handling authentication and CRUD operations for tasks and todos.
- Created LoginPage for user authentication.
- Enhanced DashboardPage to manage tasks and todos with a responsive layout.

KMKINGMAN commented:
currently working on test-units

mkurkar committed:
refactor: Replace login_view function with LoginView class-based view and update URL routing

mkurkar committed:
feat: Add delete functionality for tasks and todos with confirmation dialogs

mkurkar committed:
Implements pytest-based testing infrastructure with complete coverage of models, API endpoints, authentication, and integration workflows.

Includes factory classes for test data generation, shared fixtures for authentication and database setup, and comprehensive test cases covering CRUD operations, user permissions, data isolation, and error handling.

Establishes testing best practices with proper markers for unit, integration, and API tests to support reliable development and deployment workflows.

KMKINGMAN commented:
what i should do next is create a requirements folder 
create a Makefile (i know the main scope of it is for tracing c and cpp file) [GNU MAKEFILE DOCS](https://www.gnu.org/software/make/manual/make.html)
create an ERR Diagram for the database 
create a custom library for the django-rest-framework 
reading more in [CDRF](https://www.cdrf.co/) and [Django-Docs](https://docs.djangoproject.com/en/5.2/)

mkurkar committed:
feat: Implement user authentication flow with login and logout functionality

mkurkar committed:
feat: Refactor Todo application structure and improve code readability

- Updated admin interface for Todo and Task models.
- Enhanced app configuration with descriptive docstrings.
- Added module-level docstrings for models and serializers.
- Improved string representation methods in models.
- Refactored serializers for better clarity and organization.
- Removed obsolete test cases and cleaned up views.
- Updated settings to use the new app configuration.
- Added Pylint configuration for code quality checks.

KMKINGMAN commented:
finishing form fixing pylint errors on the code 'it's make me suffer a little bit'

mkurkar committed:
Refactor: make them use pylint

mkurkar committed:
feat: Update pylint configuration and improve code readability across models, serializers, and views

mkurkar committed:
feat: Remove pylint disable comments and add a script to validate imports and syntax across the application

mkurkar committed:
chore: Remove edx-lint generated comments and update load-plugins in pylintrc

mkurkar committed:
chore: Remove unnecessary comment from pylint configuration

KMKINGMAN commented:
fixing pylint errors

mkurkar committed:
add the resources

mkurkar committed:
refactor: make models follow flake8

mkurkar committed:
refactor: make views follow flake8

mkurkar committed:
refactor: make the tests follow flake8

mkurkar committed:
feat: add token validation and protected routes for authentication

mkurkar committed:
refactor: make it follow pep8 pylint and flake8

OmarIthawi commented on backend/app/admin.py:
```diff
@@ -0,0 +1,95 @@
+    )
+    search_fields = ('title', 'description', 'tags')
+    list_editable = ('completed',)
+    list_per_page = 25
```
keep the default, it's much more than 25 and it's better that way

```suggestion
```

OmarIthawi commented on backend/app/admin.py:
```diff
@@ -0,0 +1,95 @@
+    )
+    list_filter = ('user', 'created_at', 'updated_at')
+    search_fields = ('title', 'description', 'user__username')
+    list_per_page = 25
```
```suggestion
```

OmarIthawi commented on backend/app/admin.py:
```diff
@@ -0,0 +1,95 @@
+    readonly_fields = ('created_at', 'updated_at')
+    inlines = [TodoInline]
+
+    def todos_count(self, obj):
+        """Return the number of todos in this task."""
+        return obj.todos.count()
```
this is going to be slow, fyi

it's called N+1 query problem, perhaps not a good fit for proudction code

consider either removing this feature or use the query set itself to aggregate the `Count` in the `get_queryset`

OmarIthawi commented on backend/app/models/__init__.py:
```diff
@@ -0,0 +1,10 @@
+This module exports all model classes for the todo application.
+"""
+
+from .task_model import Task
+from .todo_model import Todo, get_default_tags
```
this app is too small to split the models into two modules

mkurkar committed:
Update backend/app/admin.py

Co-authored-by: Omar Al-Ithawi <i@omardo.com>

mkurkar committed:
Apply suggestion from @OmarIthawi

Co-authored-by: Omar Al-Ithawi <i@omardo.com>

tehreem-sadat commented on backend/app/admin.py:
```diff
@@ -0,0 +1,93 @@
+        queryset = super().get_queryset(request)
+        if not request.user.is_superuser:
+            queryset = queryset.filter(user=request.user)
+        return queryset
```
**Todo->Task.count()** translates into `SELECT COUNT(*) FROM todo WHERE task_id = <obj.id>` and as you are using it in list_display it can still cause N+1 problem as Omar mentioned in his comment.

Please explore annotate, select_related and prefetch_related and use the one that suits best here.

tehreem-sadat commented on backend/app/models/todo_model.py:
```diff
@@ -22,22 +27,31 @@ class Todo(models.Model):
-    user = models.ManyToManyField(
-        User, 
-        related_name='todos', 
+    user = models.ForeignKey(
```
Why do we need **user** foreign key in both Task and Todo model ?

tehreem-sadat commented on backend/app/serializers/task_serializer.py:
```diff
@@ -0,0 +1,113 @@
+        data = super().to_representation(instance)
+        request = self.context.get('request')
+
+        if request and request.query_params.get('include_todos') != '1':
```
Right now in the code we are fetching todos, serializes them and after all processing - in the last step throws them away if include_todos is not set.

**My suggestion  here:**
Move the query param / URL param handling (include_todos) to the view, where request-specific logic belongs. Pass include_todos into the serializer via serializer_context and then  update serilizer logic to skip adding the field from the start.

tehreem-sadat commented on backend/app/views/task_view.py:
```diff
@@ -0,0 +1,47 @@
+
+        # If user is superuser, delete all todos in this task
+        if self.request.user.is_superuser:
+            user_todos = instance.todos.all()
```
```suggestion
            user_todos = instance.todos.all()
            # Delete the todos
            for todo in user_todos:
                todo.delete()
```

instance.todos.all().delete()

mkurkar replied:
i do it

tehreem-sadat commented on backend/app/serializers/todo_serializer.py:
```diff
@@ -0,0 +1,34 @@
+        ]
+        read_only_fields = ['created_at', 'updated_at', 'user']
+
+    def create(self, validated_data):
```
Why do we need to override create here as perform_create() in the viewset already sets the user ?

mkurkar replied:
case if someone forgets to set the user in a new context
this will fix the potential mistakes

tehreem-sadat commented on backend/app/serializers/task_serializer.py:
```diff
@@ -0,0 +1,113 @@
+                # Create new todo
+                todo = Todo.objects.create(
+                    **{k: v for k, v in todo_data.items() if k != 'id'})
+                todo.user.add(self.context['request'].user)
```
Is this line working fine ? User is FK here and 'add' is mostly used for M2M. Please verify one more time.

mkurkar committed:
fix: N+1 problem

mkurkar committed:
refactor: Removes Django template views and HTML templates
feat:
Adds dotenv support for environment-based configuration management.
Moves secret key configuration to environment variables for better security practices.

mkurkar committed:
feat: Enhance TaskSerializer and TaskViewSet to conditionally include todos based on request parameters

mkurkar committed:
feat: Add getTask API method to fetch a task with optional nested todos

mkurkar committed:
fix: Update pylint plugins and improve pytest markers descriptions

mkurkar committed:
feat: Add development dependencies to requirements files

mkurkar committed:
refactor: Remove unused dashboard import and update LoginView docstring with resource links

mkurkar committed:
style: make them follow flake8

mkurkar committed:
feat: Add Tox integration for automated testing and code quality checks

mkurkar committed:
feat: Add unit tests for TaskAdmin and TaskSerializer update behavior

mkurkar committed:
refactor:

- Updated string quotes to use double quotes consistently across all files.
- Reformatted model fields and serializer fields for better readability.
- Improved comments and docstrings for clarity.
- Adjusted import order and spacing for better organization.
- Ensured consistent use of blank lines and indentation throughout the codebase.
